<a href="https://colab.research.google.com/github/jessicaromero-ctrl/Advanced-Math-for-ML/blob/main/DecisionTrees_CancerDB_py.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [22]:
from sklearn import datasets

In [23]:
from sklearn import tree
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score

In [24]:
dataset = datasets.load_breast_cancer()
print(dataset)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

In [25]:
import pandas as pd
from sklearn.datasets import load_breast_cancer

# 1. Cargar el dataset
cancer = load_breast_cancer()

# 2. Crear un DataFrame de pandas con los datos y nombres de columnas
df = pd.DataFrame(data=cancer.data, columns=cancer.feature_names)

# 3. Agregar la columna objetivo (target)
df['target'] = cancer.target

# 4. Mostrar las primeras filas
#print(df.head())
df.head

# Para ver la descripción general
#print(df.info())

<bound method NDFrame.head of      mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                   0.07871  ...          17.33           184.60      2019.0   
1                   0.05667  ...          23.41           158.80      1956.0   
2                   0.05999  ...          25.53           152.50      1709.0   
3                   0.09744  ...          26.50            98.87       567.7   
4                   0.05883  ...          16.67           152.20      1575.0   
..                      ...  ...            ...              ...         ...   
564                 0.05623  ...          26.40           166.10      2027.0   
565                 0.05533  ...          38.25           155.00      1731.0   
566                 0.05648  ...          34.12           126.70      1124.0   
567                 0.07016  ...          39.42           184.60      1821.0   
568                 0.05884  ...          30.37            59.16       268.6   

     worst smoothness  worst compactness  worst concavity  \
0             0.16220            0.66560           0.7119   
1             0.12380            0.18660           0.2416   
2             0.14440            0.42450           0.4504   
3             0.20980            0.86630           0.6869   
4             0.13740            0.20500           0.4000   
..                ...                ...              ...   
564           0.14100            0.21130           0.4107   
565           0.11660            0.19220           0.3215   
566           0.11390            0.30940           0.3403   
567           0.16500            0.86810           0.9387   
568           0.08996            0.06444           0.0000   

     worst concave points  worst symmetry  worst fractal dimension  target  
0                  0.2654          0.4601                  0.11890       0  
1                  0.1860          0.2750                  0.08902       0  
2                  0.2430          0.3613                  0.08758       0  
3                  0.2575          0.6638                  0.17300       0  
4                  0.1625    

In [26]:
print("Información en el Dataset: ")
print(dataset.keys())
print

Información en el Dataset: 
dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


<function print(*args, sep=' ', end='\n', file=None, flush=False)>

In [27]:
#print("Características del Dataset: ")
#print(dataset.DESCR())

In [28]:
X = dataset.data
y = dataset.target

In [30]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier # Added this import

#Separación de los datos
X_train, X_test, y_train, y_test = train_test_split (X, y, test_size=0.2)

#Se escalan los datos
#from sklearn.preprocessing import StandardScaler
escalar = StandardScaler()
X_train = escalar.fit_transform(X_train)
X_test = escalar.transform(X_test)

#Se define el algoritmo a utilizar
algoritmo = DecisionTreeClassifier()

#Entrenamos el modelo
algoritmo.fit(X_train, y_train);

#Se realiza la predicción
y_pred = algoritmo.predict(X_test)

#Se realiza la probabilidad
y_proba = algoritmo.predict_proba(X_test)[:, 1]

#Se verifica con la matriz de confusión
from sklearn.metrics import confusion_matrix
matriz = confusion_matrix(y_test, y_pred)
print("Matriz de confusión: ")
print(matriz)

#Se calcula la precisión del modelo
from sklearn.metrics import precision_score

precision = precision_score(y_test,y_pred)
print("Precisión del Modelo: ")
print(precision)

#Calculamos la exactitud del modelo
from sklearn.metrics import accuracy_score

exactitud = accuracy_score(y_test,y_pred)
print("Exactitud del Modelo: ")
print(exactitud)

#Calculamos la sensibilidad del modelo
from sklearn.metrics import recall_score

sensibilidad = recall_score(y_test,y_pred)
print("Sensibilidad del Modelo: ")
print(sensibilidad)

#Calculamos el puntaje F1 del modelo
from sklearn.metrics import f1_score

puntajeF1 = f1_score(y_test,y_pred)
print("Puntaje F1 del Modelo: ")
print(puntajeF1)

#Calculamos la curva ROC-AUC del modelo
from sklearn.metrics import roc_auc_score

#roc_auc = roc_auc_score(y_test,y_pred)
roc_auc = roc_auc_score(y_test, y_proba)
print("Curva ROC - AUC del Modelo: ")
print(roc_auc)

#print("Precisión del Modelo: ", precision)
#print("Exactitud del Modelo: ",exactitud)
#print("Sensibilidad del Modelo: ",sensibilidad)
#print("Puntaje F1 del Modelo: ", puntajeF1)
#print("Curva ROC - AUC del Modelo: ", roc_auc)

Matriz de confusión: 
[[44  3]
 [ 5 62]]
Precisión del Modelo: 
0.9538461538461539
Exactitud del Modelo: 
0.9298245614035088
Sensibilidad del Modelo: 
0.9253731343283582
Puntaje F1 del Modelo: 
0.9393939393939394
Curva ROC - AUC del Modelo: 
0.9307716735471578


In [ ]:
import matplotlib.pyplot as plt
from sklearn.metrics import roc_curve, auc

# 1. Obtener probabilidades de la clase positiva ***PUEDE OMITIRSE PORQUE SE OBTUVO PREVIAMENTE***
#y_proba = algoritmo.predict_proba(X_test)[:, 1]

# 2. Calcular FPR, TPR y umbrales
fpr, tpr, thresholds = roc_curve(y_test, y_proba)

# 3. Calcular el área bajo la curva (AUC)
roc_auc = auc(fpr, tpr)

# 4. Graficar
plt.figure()
plt.plot(fpr, tpr, label=f"Curva ROC (AUC = {roc_auc:.3f})")
plt.plot([0, 1], [0, 1], linestyle="--", label="Clasificador aleatorio")

plt.xlabel("Tasa de Falsos Positivos (FPR)")
plt.ylabel("Tasa de Verdaderos Positivos (TPR)")
plt.title("Curva ROC - Regresión Logística")
plt.legend(loc="lower right")
plt.grid(True)
plt.show()

## Resultados:
Matriz de confusión:

[[44  3]
 [ 5 62]]

This matrix indicates that the model **correctly classified 44 negative cases (true negatives) and 62 positive cases (true positives).** It misclassified 3 negative cases as positive (false positives) and 5 positive cases as negative (false negatives).

Precisión del Modelo (Precision):** 0.954 This means that when the model predicted a positive outcome, it was correct approximately 95.4% of the time.**

Exactitud del Modelo (Accuracy): **0.930 The model correctly classified about 93.0% of all the samples.**

Sensibilidad del Modelo (Recall): **0.925 This indicates that the model correctly identified 92.5% of all actual positive cases.**

Puntaje F1 del Modelo (F1 Score):** 0.939 The F1 Score provides a balance between precision and recall, with a value of 0.939 suggesting a good overall performance, especially useful when there might be an uneven class distribution.**

Curva ROC - AUC del Modelo (ROC AUC Score):** 0.931 An AUC score of 0.931 suggests that the model has a very good ability to distinguish between the positive and negative classes.**